rust/src/pgsql/pgsql.rs

```diff
@@ -0,0 +1,853 @@
+                    return AppLayerResult::err();
+                }
+            }
+        }
+
+        // All input was fully consumed.
+        return AppLayerResult::ok();
+    }
+
+    fn tx_iterator(
+        &mut self, min_tx_id: u64, state: &mut u64,
+    ) -> Option<(&PgsqlTransaction, u64, bool)> {
+        let mut index = *state as usize;
+        let len = self.transactions.len();
+
+        while index < len {
+            let tx = &self.transactions[index];
+            if tx.tx_id < min_tx_id + 1 {
+                index += 1;
+                continue;
+            }
+            *state = index as u64;
+            return Some((tx, tx.tx_id - 1, (len - index) > 1));
+        }
+
+        return None;
+    }
+
+    fn on_request_gap(&mut self, _size: u32) {
+        self.request_gap = true;
+    }
+
+    fn on_response_gap(&mut self, _size: u32) {
+        self.response_gap = true;
+    }
+}
+
+/// Probe for a valid PostgreSQL request
+///
+/// PGSQL messages don't have a header per se, so we parse the slice for an ok()
+fn probe_ts(input: &[u8]) -> bool {
+    SCLogDebug!("We are in probe_ts");
+    parser::parse_request(input).is_ok()
+}
+
+/// Probe for a valid PostgreSQL response
+///
+/// PGSQL messages don't have a header per se, so we parse the slice for an ok()
+fn probe_tc(input: &[u8]) -> bool {
+    if parser::pgsql_parse_response(input).is_ok() || parser::parse_ssl_response(input).is_ok() {
+        return true;
+    }
+    SCLogDebug!("probe_tc is false");
+    false
+}
+
+// C exports.
+
+export_tx_get_detect_state!(rs_pgsql_tx_get_detect_state, PgsqlTransaction);
+export_tx_set_detect_state!(rs_pgsql_tx_set_detect_state, PgsqlTransaction);
+
+/// C entry point for a probing parser.
+#[no_mangle]
+pub unsafe extern "C" fn rs_pgsql_probing_parser_ts(
+    _flow: *const Flow, _direction: u8, input: *const u8, input_len: u32, _rdir: *mut u8,
+) -> AppProto {
+    if input_len >= 1 && input != std::ptr::null_mut() {
+        let slice: &[u8];
+        slice = build_slice!(input, input_len as usize);
+        if probe_ts(slice) {
+            return ALPROTO_PGSQL;
+        }
+    }
+    return ALPROTO_UNKNOWN;
+}
+
+/// C entry point for a probing parser.
+#[no_mangle]
+pub unsafe extern "C" fn rs_pgsql_probing_parser_tc(
+    _flow: *const Flow, _direction: u8, input: *const u8, input_len: u32, _rdir: *mut u8,
+) -> AppProto {
+    if input_len >= 1 && input != std::ptr::null_mut() {
+        let slice: &[u8];
+        slice = build_slice!(input, input_len as usize);
+        if probe_tc(slice) {
+            return ALPROTO_PGSQL;
+        }
+    }
+    return ALPROTO_UNKNOWN;
+}
+
+#[no_mangle]
+pub extern "C" fn rs_pgsql_state_new(
+    _orig_state: *mut std::os::raw::c_void, _orig_proto: AppProto,
+) -> *mut std::os::raw::c_void {
+    let state = PgsqlState::new();
+    let boxed = Box::new(state);
+    return Box::into_raw(boxed) as *mut _;
```

Comment:
Why aren't we explicitly defining the raw void pointer here and using a `_`?

Reply:
I followed what was done in other parsers, as this was a new approach to me, I didn't know what was the best way...

Reply:
If a type is required, but can easily be inferred you can just use a _.  So this is equivalent to 
```
return Box::into_raw(boxed) as *mut std::os::raw::c_void;
```
but using `_` was a litte more convenient for cut and paste when doing this on mass when a few of the types were different.

Reply:
I see. Thanks for the explanation... Is it preferable to explicitly define them, here, then?

Reply:
No preference either way.